dali/kernels/signal/resampling_cpu.cc

```diff
@@ -0,0 +1,282 @@
+  float32x2_t c3 = vld1_f32(&lookup[idx[3]]);
+  float32x4x2_t w = vuzpq_f32(vcombine_f32(c0, c1), vcombine_f32(c2, c3));
+  float32x4_t curr = w.val[0];
+  float32x4_t next = w.val[1];
+  return vfmaq_f32(curr, di, vsubq_f32(next, curr));
+}
+
+inline float32x4_t vsetq_f32(float x0, float x1, float x2, float x3) {
+  float32x4_t x;
+  x = vdupq_n_f32(x0);
+  x = vsetq_lane_f32(x1, x, 1);
+  x = vsetq_lane_f32(x2, x, 2);
+  x = vsetq_lane_f32(x3, x, 3);
+  return x;
+}
+
+inline float filter_vec(ResamplingWindow &window, int &i_ref, float in_pos, int i1,
+                        const float *in) {
+  const float32x4_t _0123 = vsetq_f32(0, 1, 2, 3);
+  float32x4_t f4 = vdupq_n_f32(0);
+
+  int i = i_ref;
+  float32x4_t x4 = vaddq_f32(vdupq_n_f32(i - in_pos), _0123);
+
+  for (; i + 3 < i1; i += 4) {
+    float32x4_t w4 = evaluate(window, x4);
+    f4 = vfmaq_f32(f4, vld1q_f32(in + i), w4);
+    x4 = vaddq_f32(x4, vdupq_n_f32(4));
+  }
+  // Sum elements in f4
+  float32x2_t f2 = vpadd_f32(vget_low_f32(f4), vget_high_f32(f4));
+  f2 = vpadd_f32(f2, f2);
+  i_ref = i;
+  return vget_lane_f32(f2, 0);
+}
+
+#elif defined(__SSE2__)
+
+inline __m128 evaluate(const ResamplingWindow &w, __m128 x) {
+  __m128 fi = _mm_add_ps(x * _mm_set1_ps(w.scale), _mm_set1_ps(w.center));
+  __m128i i = _mm_cvttps_epi32(fi);
+  __m128 fifloor = _mm_cvtepi32_ps(i);
+  __m128 di = _mm_sub_ps(fi, fifloor);
+  int idx[4];
+  _mm_storeu_si128(reinterpret_cast<__m128i *>(idx), i);
+  __m128 curr = _mm_setr_ps(w.lookup[idx[0]], w.lookup[idx[1]], w.lookup[idx[2]], w.lookup[idx[3]]);
+  __m128 next =
+      _mm_setr_ps(w.lookup[idx[0] + 1], w.lookup[idx[1] + 1], w.lookup[idx[2] + 1], w.lookup[idx[3] + 1]);
+  return _mm_add_ps(curr, _mm_mul_ps(di, _mm_sub_ps(next, curr)));
+}
+
+inline float filter_vec(ResamplingWindow &window, int &i_ref, float in_pos, int i1,
+                        const float *in) {
+  __m128 f4 = _mm_setzero_ps();
+  int i = i_ref;
+  __m128 x4 = _mm_setr_ps(i - in_pos, i + 1 - in_pos, i + 2 - in_pos, i + 3 - in_pos);
+  for (; i + 3 < i1; i += 4) {
+    __m128 w4 = evaluate(window, x4);
+
+    f4 = _mm_add_ps(f4, _mm_mul_ps(_mm_loadu_ps(in + i), w4));
+    x4 = _mm_add_ps(x4, _mm_set1_ps(4));
+  }
+  i_ref = i;
+
+  // Sum elements in f4
+  f4 = _mm_add_ps(f4, _mm_shuffle_ps(f4, f4, _MM_SHUFFLE(1, 0, 3, 2)));
+  f4 = _mm_add_ps(f4, _mm_shuffle_ps(f4, f4, _MM_SHUFFLE(0, 1, 0, 1)));
+  return _mm_cvtss_f32(f4);
+}
+
+#else
+
+inline float filter_vec(ResamplingWindow &, int &, float, int, const float *) {
+  return 0;
+}
+
+#endif
+
+/**
+ * @brief Resample single-channel signal and convert to Out
+ *
+ * Calculates a range of resampled signal.
+ * The function can seamlessly resample the input and produce the result in chunks.
+ * To reuse memory and still simulate chunk processing, adjust the in/out pointers.
+ */
+template <typename Out>
+void ResampleCPUImpl(ResamplingWindow &window, Out *__restrict__ out, int64_t out_begin,
```

Comment:
```suggestion
void ResampleCPUImpl(ResamplingWindow window, Out *__restrict__ out, int64_t out_begin,
```
or
```suggestion
void ResampleCPUImpl(const ResamplingWindow &window, Out *__restrict__ out, int64_t out_begin,
```
I think passing by value can facilitate optimization.